internal/guest/oom/oom.go

```diff
@@ -0,0 +1,80 @@
+// Package oom providers wrappers and support to leverage the OOM watcher functionality
+// in github.com/containerd/containerd/pkg/oom
+package oom
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/guest/prot"
+	log "github.com/Microsoft/hcsshim/internal/log"
+	eventstypes "github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/containerd/runtime"
+	"github.com/containerd/containerd/runtime/v2/shim"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+)
+
+var emptyActivityID = (uuid.UUID{}).String()
+
+// wraps `bridge.PublishNotification` to conform to `shim.Publish`, so that
+// it can be used by containerd's `oom.Watcher`
+type BridgePublisherFunc func(*prot.ContainerNotification)
+
+var _ shim.Publisher = BridgePublisherFunc(func(_ *prot.ContainerNotification) {})
+
+func (bp BridgePublisherFunc) Close() error {
+	// noop
+	return nil
+}
+
+func (bp BridgePublisherFunc) Publish(ctx context.Context, topic string, event events.Event) error {
+	n, err := mapEventToNotification(topic, event)
+
+	if err != nil {
+		return errors.Wrapf(err, "could not handle topic %q and event %+v", topic, event)
+	}
+
+	log.G(ctx).WithField("event", fmt.Sprintf("%+v", event)).WithField("eventTopic", topic).Debug("publishing event over notification channel")
+
+	bp(n)
+	return nil
+}
+
+// may want to expand this to other events in the future
+func mapEventToNotification(topic string, event events.Event) (*prot.ContainerNotification, error) {
+	if topic != runtime.TaskOOMEventTopic {
+		return nil, fmt.Errorf("unsupported event topic %q", topic)
+	}
+
+	e, ok := event.(*eventstypes.TaskOOM)
+	if !ok {
+		return nil, fmt.Errorf("unsupported event type %T", e)
+	}
+
+	info := make(map[string]interface{})
+	// add time information to event, will be belated, but still close enough
+	info["time"] = time.Now()
+	// err != nil implies infoB == nil, which is in an appropriate fall back
+	infoB, _ := json.Marshal(info)
+
+	// TODO: add logging here
```

Comment:
I'd clean up some of the TODO's around before we check in

Reply:
todo was outdated (logging is handled in `Publish`, and by bridge
Adding in timestamp was also unnecessary and unused, so removed that as well

Reply:
Sweet! Resolving